src/psyclone/psyir/transformations/hoist_local_arrays_trans.py

```diff
@@ -0,0 +1,291 @@
+            (that is not a FileContainer).
+        :raises TransformationError: if any symbols corresponding to local \
+            arrays have a tag that already exists in the table of the parent \
+            Container.
+        :raises TransformationError: if any symbols corresponding to local \
+            arrays appear within any CodeBlocks inside the Routine.
+
+        '''
+        super().validate(node, options=options)
+
+        # The node should be a Routine.
+        if not isinstance(node, Routine):
+            raise TransformationError(
+                f"The target of the HoistLocalArraysTrans transformation "
+                f"should be a Routine but found '{type(node).__name__}'.")
+
+        if node.is_program:
+            # We silently ignore routines that are programs - this
+            # transformation will do nothing.
+            return
+
+        # The Routine must be within a Container (otherwise we have nowhere
+        # to hoist any array declarations to).
+        container = node.ancestor(Container)
+        if not container:
+            raise TransformationError(
+                f"The supplied routine '{node.name}' should "
+                f"be within a Container but none was found.")
+        if isinstance(container, FileContainer):
+            raise TransformationError(
+                f"The supplied routine '{node.name}' should be within a "
+                f"Container but the enclosing container is a "
+                f"FileContainer (named '{container.name}').")
+
+        # Check for clashing tags in the container scope.
+        auto_arrays = self._get_local_arrays(node)
+        tags_dict = node.symbol_table.get_reverse_tags_dict()
+        cont_tags_dict = container.symbol_table.tags_dict
+        for sym in auto_arrays:
+            tag = tags_dict.get(sym)
+            if tag in cont_tags_dict:
+                raise TransformationError(
+                    f"The supplied routine '{node.name}' contains a local "
+                    f"array '{sym.name}' with tag '{tag}' but this tag is "
+                    f"also present in the symbol table of the parent "
+                    f"Container (associated with variable "
+                    f"'{cont_tags_dict[tag].name}').")
+
+        # Check that none of the arrays to be hoisted are accessed with a
+        # CodeBlock (as they may get renamed as part of the transformation).
+        array_names = set(sym.name for sym in auto_arrays)
+        cblocks = node.walk(CodeBlock)
+        for cblock in cblocks:
+            cblock_names = set(cblock.get_symbol_names())
+            if cblock_names.intersection(array_names):
+                cblock_txt = ''.join(str(fpn) for fpn in cblock.get_ast_nodes)
+                raise TransformationError(
+                    f"The supplied routine '{node.name}' contains a local "
+                    f"array '{sym.name}' which is accessed within a CodeBlock "
```

Comment:
pylint says: undefined-loop-variable: Using possibly undefined loop variable 'sym'

And btw is really sym.name (the last sym in auto_arrays) what you want here? Or should the `cblock_names.intersection(array_names)` be the list of array names to print?

Reply:
Also, do we need to fail? Could this be inside the `_get_local_arrays` and just filter out the ones inside CodeBlocks but continue hoisting the others? Printing a log message would still be useful but we don't do this yet right?

Anyway, its just a suggestion, I leave it up to you to decide what's best.

Reply:
You and sympy are both right - that code was incorrect. Anyway, I like your suggestion of not failing so have moved this functionality to `_get_local_arrays()` and repurposed the existing test.